Add an empty core API with semi-stubs.

dmitvitalii opened the pull request:
Add methods to register/unregister module, to enable and disable it.

TODO: find a solution on working wit a lifetime information.

dmitvitalii committed:
Add an empty core API with semi-stubs.

Add methods to register/unregister module, to enable and disable it.

TODO: find a solution on working wit a lifetime information.

riseremi commented on src/core.js:
```diff
@@ -0,0 +1,38 @@
+// TODO: non-working code. Implement after accepting an API.
+
+var modules = {};
```
For mutable variables, prefer `let` over `var`.

dmitvitalii replied:
Done. Made it `const`.

riseremi commented on src/core.js:
```diff
@@ -0,0 +1,38 @@
+    if (modules[command].enabled) {
+        modules[command].onDisable();
+    }
+    delete modules[command];
```
Using `delete` looks odd to me. Consider using ES6 `Map` which allows to delete entries.

dmitvitalii replied:
Done.

riseremi commented on src/core.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+function handleCommand(command, args) {
+    if (modules[command].onReceive === undefined) {
```
You can reject modules without `onReceive` handler in `registerModule` function. What's the point of module that does nothing?

dmitvitalii replied:
Done.

riseremi commented on src/core.js:
```diff
@@ -0,0 +1,38 @@
+    if (modules[command].onReceive === undefined) {
+        return -1; // throw NoOnReceiveMethodException()
+    }
+    if (!modules[command].enabled) {
```
Can we ignore disabled modules here?

dmitvitalii replied:
To avoid an accidental removing of the function from the object, I'm considering to leave checking of the onReceive function availability, even when we won't allow adding a module without it.

riseremi reviewed (changes requested):
As a general rule: you should indent code with two spaces.

dmitvitalii committed:
Review fixes.

[*] modules became a Map object;
[*] rename modules to MODULES and make it const instead of var;
[*] reject modules without onReceive() function;
[*] add JSDocs.

riseremi commented on src/core.js:
```diff
@@ -0,0 +1,86 @@
+ * @type {Map.<Object>}
+ * @constant
+ */
+const MODULES = new Map();
```
I think this should be lowercase, since this is mutable variable anyway. What do you think?

dmitvitalii replied:
It depends on a codestyle we're using. I find an information about the mutable constants situation only in a Google Guileline, but there are several options:
* [idiomatic.js](https://github.com/rwaldron/idiomatic.js/tree/master/translations/ru_RU): Symbolic constants always UPPER_CASE;
* [dojo](https://dojotoolkit.org/reference-guide/1.10/developer/styleguide.html): Constants either CamelCase (like ClassName) or UPPER_CASE;
* [jQuery](http://contribute.jquery.org/style-guide/js/#naming-conventions): No info about `const`, always mixedCamelCase;
* [Google](https://google.github.io/styleguide/javascriptguide.xml): Use UPPER_CASE for constant _values_, don't use `const` keyword because of `IE`, use `@const` in JSDoc instead;
* [Airbnb](https://github.com/airbnb/javascript): Always use mixedCamelCase, even with `const` vars.

Personaly, I didn't want to make `MODULES` uppercase, so I can suggest to use Google's styleguide. If we approve it (or some other styleguide, or even develop smth different), I will make it lowercase and use "mutable const is lowercase" as a rule of thumb.

dmitvitalii replied:
Decided to use Airbnb codestyle.

riseremi reviewed (commented):
Looks nice to me overall!

Commit:
Lowercase the modules map const name.
